File: apps/backend-gateway/src/interfaces/middleware/request_logging.py
"""
StadiumOS AI — Request Logging Middleware.

Intercepts requests to log route info, compute execution time, and inject
unique request IDs for transaction tracing.
"""

import time
import uuid
from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
import structlog


class RequestLoggingMiddleware(BaseHTTPMiddleware):
    """
    Middleware that attaches trace contexts and monitors request metrics.

    Steps:
        1. Extracts or generates a trace identifier ('X-Request-ID').
        2. Binds trace details into structlog's thread-local/coroutine-local context vars.
        3. Measures processing latency.
        4. Logs start and completion events.
        5. Returns the request ID header to the client.
    """

    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        # Extract existing Request ID from header or generate a new UUID
        request_id = request.headers.get("x-request-id")
        if not request_id:
            request_id = str(uuid.uuid4())

        # Bind the request context for all log statements executed during this request
        structlog.contextvars.clear_contextvars()
        structlog.contextvars.bind_contextvars(
            request_id=request_id,
            method=request.method,
            path=request.url.path,
            client_ip=request.client.host if request.client else "unknown",
        )

        logger = structlog.get_logger("request_tracker")
        start_time = time.perf_counter()

        await logger.ainfo("request_started")

        try:
            response = await call_next(request)
        except Exception as exc:
            duration = (time.perf_counter() - start_time) * 1000
            await logger.aerror(
                "request_failed",
                duration_ms=round(duration, 2),
                error=str(exc),
                exc_info=True,
            )
            # Re-raise so the global exception handler can translate it into a JSON response
            raise exc

        duration = (time.perf_counter() - start_time) * 1000
        
        # Log request completion with duration and status code
        await logger.ainfo(
            "request_completed",
            status_code=response.status_code,
            duration_ms=round(duration, 2),
        )

        # Attach request ID to response headers for debugging tracking
        response.headers["X-Request-ID"] = request_id
        return response

File: apps/backend-gateway/src/interfaces/middleware/requestLogging.ts
/**
 * StadiumOS AI — Request Logging Middleware.
 * 
 * Intercepts requests to inject correlation transaction IDs, log route entries,
 * and track request latency durations.
 */

import { Request, Response, NextFunction } from "express";
import { v4 as uuidv4 } from "uuid";
import { logger } from "../../infrastructure/logging/logger";

export interface LoggedRequest extends Request {
  requestId?: string;
  startTime?: [number, number];
}

export function requestLoggingMiddleware(
  req: LoggedRequest,
  res: Response,
  next: NextFunction
): void {
  // Extract or generate unique request ID
  const requestId = (req.headers["x-request-id"] as string) || uuidv4();
  req.requestId = requestId;
  req.startTime = process.hrtime();

  // Attach request ID to response header
  res.setHeader("X-Request-ID", requestId);

  const clientIp = req.ip || req.socket.remoteAddress || "unknown";

  // Bind trace info to local logger context
  const childLogger = logger.child({
    requestId,
    method: req.method,
    path: req.url,
    clientIp
  });

  childLogger.info("Request started");

  res.on("finish", () => {
    const elapsed = process.hrtime(req.startTime);
    // Convert hrtime to duration in milliseconds
    const durationMs = (elapsed[0] * 1000 + elapsed[1] / 1e6).toFixed(2);
    
    childLogger.info(
      {
        statusCode: res.statusCode,
        durationMs: parseFloat(durationMs)
      },
      "Request completed"
    );
  });

  next();
}
